// angular-in-memory-web-api/esm5/http-client-backend.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Inject, Injectable, Optional } from '@angular/core';
import { HttpHeaders, HttpParams, HttpResponse, HttpXhrBackend, XhrFactory } from '@angular/common/http';
import { map } from 'rxjs/operators';
import { STATUS } from './http-status-codes';
import { InMemoryBackendConfig, InMemoryBackendConfigArgs, InMemoryDbService } from './interfaces';
import { BackendService } from './backend.service';
/**
 * For Angular `HttpClient` simulate the behavior of a RESTy web api
 * backed by the simple in-memory data store provided by the injected `InMemoryDbService`.
 * Conforms mostly to behavior described here:
 * http://www.restapitutorial.com/lessons/httpmethods.html
 *
 * ### Usage
 *
 * Create an in-memory data store class that implements `InMemoryDbService`.
 * Call `config` static method with this service class and optional configuration object:
 * ```
 * // other imports
 * import { HttpClientModule } from '\@angular/common/http';
 * import { HttpClientInMemoryWebApiModule } from 'angular-in-memory-web-api';
 *
 * import { InMemHeroService, inMemConfig } from '../api/in-memory-hero.service';
 * \@NgModule({
 *  imports: [
 *    HttpModule,
 *    HttpClientInMemoryWebApiModule.forRoot(InMemHeroService, inMemConfig),
 *    ...
 *  ],
 *  ...
 * })
 * export class AppModule { ... }
 * ```
 */
var HttpClientBackendService = /** @class */ (function (_super) {
    tslib_1.__extends(HttpClientBackendService, _super);
    function HttpClientBackendService(inMemDbService, config, xhrFactory) {
        var _this = _super.call(this, inMemDbService, config) || this;
        _this.xhrFactory = xhrFactory;
        return _this;
    }
    /**
     * @param {?} req
     * @return {?}
     */
    HttpClientBackendService.prototype.handle = /**
     * @param {?} req
     * @return {?}
     */
    function (req) {
        try {
            return this.handleRequest(req);
        }
        catch (error) {
            /** @type {?} */
            var err = error.message || error;
            /** @type {?} */
            var resOptions_1 = this.createErrorResponseOptions(req.url, STATUS.INTERNAL_SERVER_ERROR, "" + err);
            return this.createResponse$((/**
             * @return {?}
             */
            function () { return resOptions_1; }));
        }
    };
    ////  protected overrides /////
    ////  protected overrides /////
    /**
     * @protected
     * @param {?} req
     * @return {?}
     */
    HttpClientBackendService.prototype.getJsonBody = 
    ////  protected overrides /////
    /**
     * @protected
     * @param {?} req
     * @return {?}
     */
    function (req) {
        return req.body;
    };
    /**
     * @protected
     * @param {?} req
     * @return {?}
     */
    HttpClientBackendService.prototype.getRequestMethod = /**
     * @protected
     * @param {?} req
     * @return {?}
     */
    function (req) {
        return (req.method || 'get').toLowerCase();
    };
    /**
     * @protected
     * @param {?} headers
     * @return {?}
     */
    HttpClientBackendService.prototype.createHeaders = /**
     * @protected
     * @param {?} headers
     * @return {?}
     */
    function (headers) {
        return new HttpHeaders(headers);
    };
    /**
     * @protected
     * @param {?} search
     * @return {?}
     */
    HttpClientBackendService.prototype.createQueryMap = /**
     * @protected
     * @param {?} search
     * @return {?}
     */
    function (search) {
        /** @type {?} */
        var map = new Map();
        if (search) {
            /** @type {?} */
            var params_1 = new HttpParams({ fromString: search });
            params_1.keys().forEach((/**
             * @param {?} p
             * @return {?}
             */
            function (p) { return map.set(p, params_1.getAll(p)); }));
        }
        return map;
    };
    /**
     * @protected
     * @param {?} resOptions$
     * @return {?}
     */
    HttpClientBackendService.prototype.createResponse$fromResponseOptions$ = /**
     * @protected
     * @param {?} resOptions$
     * @return {?}
     */
    function (resOptions$) {
        return resOptions$.pipe(map((/**
         * @param {?} opts
         * @return {?}
         */
        function (opts) { return new HttpResponse(opts); })));
    };
    /**
     * @protected
     * @return {?}
     */
    HttpClientBackendService.prototype.createPassThruBackend = /**
     * @protected
     * @return {?}
     */
    function () {
        try {
            return new HttpXhrBackend(this.xhrFactory);
        }
        catch (ex) {
            ex.message = 'Cannot create passThru404 backend; ' + (ex.message || '');
            throw ex;
        }
    };
    HttpClientBackendService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    HttpClientBackendService.ctorParameters = function () { return [
        { type: InMemoryDbService },
        { type: InMemoryBackendConfigArgs, decorators: [{ type: Inject, args: [InMemoryBackendConfig,] }, { type: Optional }] },
        { type: XhrFactory }
    ]; };
    return HttpClientBackendService;
}(BackendService));
export { HttpClientBackendService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    HttpClientBackendService.prototype.xhrFactory;
}
